VariableKernel calls into scattered C++ api

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44158 VariableKernel calls into scattered C++ api**
* #44087 Add faithful C++ API

Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

codegen diff: P143616493

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

smessmer committed:
VariableKernel calls into scattered C++ api

Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

ezyang commented on tools/autograd/gen_autograd.py:
```diff
@@ -128,6 +128,8 @@ def load_aten_declarations(path):
         for arg in declaration['arguments']:
             arg['simple_type'] = get_simple_type(arg)
+        for arg in declaration['schema_order_arguments']:
+            arg['simple_type'] = get_simple_type(arg)
```
What's going on here?

smessmer replied:
`simple_type` is needed downstream for a few things and previously, only the args in `arguments` got it attached, now I'm also attaching it to the args listed in `schema_order_arguments`

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -1259,11 +1261,9 @@ def requires_unpack(arg):
-                body.append(LEGACY_UNPACK_OPTIONS.substitute(arg_name=arg['name']))
+            assert declaration['use_c10_dispatcher'] == 'with_codegenerated_unboxing_wrapper', \
+                "VariableKernel shouldn't take TensorOptions if the op is c10-full"
+            body.append(LEGACY_UNPACK_OPTIONS.substitute(arg_name=arg['name']))
```
Probably time to rename `LEGACY_UNPACK_OPTIONS`?

smessmer replied:
renamed it to `LEGACY_WRAP_OPTIONS`. It's still legacy since it's only for non-c10-full ops. Do you have a better idea?

ezyang reviewed (approved):
If possible, if you can post before and after codegen, that would be v helpful. Thanks!

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer commented:
> If possible, if you can post before and after codegen, that would be v helpful. Thanks!

@ezyang I added a link to the codegen diff in the description

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

codegen diff: P143616493

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

codegen diff: P143616493

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

codegen diff: P143616493

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

codegen diff: P143616493

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

codegen diff: P143616493

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]

smessmer committed:
Update on "VariableKernel calls into scattered C++ api"


Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering

codegen diff: P143616493

Differential Revision: [D23512538](https://our.internmc.facebook.com/intern/diff/D23512538/)

[ghstack-poisoned]
